WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,753 @@
+         stErrArgs.uiErrorNum = D_COMRV_CRC_CHECK_ERR;                           \
+         stErrArgs.uiToken    = unToken.uiValue;                                 \
+         comrvErrorHook(&stErrArgs);                                             \
+      }
+#else
+#define M_COMRV_VERIFY_CRC(pAddressToCalc, usMemSizeInBytes, uiExpectedResult)
+#endif /* D_COMRV_CRC */
 
+#ifdef D_COMRV_DEBUG
+#define M_COMRV_ASSERT(conditionMet) if (conditionMet) \
+                                     { while(1);}
+#else
+#define M_COMRV_ASSERT(conditionMet)
+#endif /* D_COMRV_DEBUG */
 /**
 * types
 */
 
 /**
 * local prototypes
 */
+void*        comrvMemset                     (void* pMemory, s32_t siVal, u32_t uiSizeInDwords);
+static u08_t comrvGetEvictionCandidates      (u08_t ucRequestedEvictionSize, u08_t* pEvictCandidatesList);
+static void  comrvUpdateCacheEvectionParams  (u08_t ucEntryIndex);
+static u16_t comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t unToken);
 
 /**
 * external prototypes
 */
+/* main comrv entry function - all overlay functions are invoked
+   through this function (the address of comrvEntry() is set in reg t6) */
+extern void  comrvEntry               (void);
+
+/* user hook functions - user application must implement the following functions */
+extern void  comrvErrorHook           (const comrvErrorArgs_t* pErrorArgs);
+extern void  comrvMemcpyHook          (void* pDest, void* pSrc, u32_t uiSizeInBytes);
+extern void* comrvLoadOvlayGroupHook  (comrvLoadArgs_t* pLoadArgs);
+#ifdef D_COMRV_CRC
+extern u32_t comrvCrcCalcHook         (const void* pAddress, u16_t usMemSizeInBytes, u32_t uiExpectedResult);
+#endif /* D_COMRV_CRC */
+#ifdef D_COMRV_FW_INSTRUMENTATION
+extern void  comrvInstrumentationHook (const comrvInstrumentationArgs_t* pInstArgs);
+#endif /* D_COMRV_FW_INSTRUMENTATION */
 
 /**
 * global variables
 */
+/* global comrv control block */
+static comrvCB_t         g_stComrvCB;
+/* global comrv stack pool */
+static comrvStackFrame_t g_stComrvStackPool[D_COMRV_CALL_STACK_DEPTH];
+
+// TODO: ronen - put pointer to tables in comrvCB_t?
+u16_t *pOverlayOffsetTable = (u16_t*)&overlayOffsetTable;
+#ifdef D_COMRV_MULTI_GROUP_SUPPORT
+comrvOverlayToken_t *pOverlayMultiGroupTokensTable = (comrvOverlayToken_t*)&overlayMultiGroupTokensTable;
+#endif /* D_COMRV_MULTI_GROUP_SUPPORT */
+
+/**
+* COM-RV initialization function
+*
+* @param  pInitParams - initialization parameters
+*
+* @return none
+*/
+void comrvInit(comrvInitArgs_t* pInitArgs)
+{
+   comrvCacheEntry_t *pCacheEntry;
+   u08_t              ucIndex;
+   void*              pBaseAddress = pInitArgs->pCacheMemoeyAddress;
+   comrvStackFrame_t* pStackPool   = g_stComrvStackPool;
+#ifdef D_COMRV_VERIFY_INIT_ARGS
+   comrvErrorArgs_t   stErrArgs;
+#endif /* D_COMRV_VERIFY_INIT_ARGS */
+
+#ifdef D_COMRV_VERIFY_INIT_ARGS
+   /* verify input parameters */
+   if ((!pInitArgs) || (pInitArgs->uiCacheSizeInBytes % D_COMRV_OVL_GROUP_SIZE_MIN))
```

Comment:
add check if input size is aligned to defined size

Reply:
done in 1e8cc05